### PR TITLE
Garter carriage detection from the left hall sensor

### PR DESCRIPTION
**Pull request by clholgat**
This works reliably. I can't test the right hall sensor because my machine has hardware issues but I can take a stab at it if y'all want.

There may be a one needle alignment issue but I'll follow up and correct that later.

### CODE REVIEW

**Comment by clholgat** [comment]
Annd the tests are broken, yea that makes sense. I'll get them fixed this weekend.

**Comment by clholgat** [comment]
I need to tweak the build to get the tests working on my mac. Closing for now until I get everything working again.

**Comment by clholgat** [comment]
Got the tests fixed!

File: test/test_encoders.cpp [hunk 1]
```diff
@@ -155,8 +155,6 @@ TEST_F(EncodersTest, test_encA_rising_in_front_G_carriage) {
   // In front of Left Hall Sensor
   EXPECT_CALL(*arduinoMock, analogRead(EOL_PIN_L))
       .WillOnce(Return(FILTER_L_MIN[encoders->getMachineType()] - 1));
-  // BeltShift is regular
-  EXPECT_CALL(*arduinoMock, digitalRead(ENC_PIN_C)).WillOnce(Return(true));
```
Comment (clholgat) [inline test/test_encoders.cpp:159]: Belt shift is only read when the first magnet passes the hall sensor

**Comment by clholgat** [comment]
Looks like Code coverage failed in the automated tests. 

(ERROR) failed minimum line coverage (got 0.0%, minimum 100.0%)

Is that configured correctly?

**Comment by chris007de** [comment]
> Looks like Code coverage failed in the automated tests.
> 
> (ERROR) failed minimum line coverage (got 0.0%, minimum 100.0%)
> 
> Is that configured correctly?

As it reports no branches and lines at all, I assume there is something wrong with the exclude patterns in GCOV_ARGS.
Some time ago it seemed to work, but it has already been broken when you contributed last time: https://github.com/AllYarnsAreBeautiful/ayab-firmware/commit/4c9236eeeb89bf4c965e5390a4c96e148d6c9161

So I would put this on the ToDo list and do the merge anyway.

**Pull request merged**